### PR TITLE
Let all SolverCG constructors take MPI_Comm.

### PR DESCRIPTION
**Pull request by marcfehling**
Part of #14426.

We can use the same parameters to construct any kind of `SolverCG` object.

PETScWrappers::SolverCG
https://github.com/dealii/dealii/blob/7e6198a4cd5e2cb9435fd7cadb2eb68bcfdf7874/include/deal.II/lac/petsc_solver.h#L404-L406

**NEW** TrilinosWrappers::SolverCG
```
    SolverCG(SolverControl &       cn,
             const MPI_Comm &      mpi_communicator,
             const AdditionalData &data = AdditionalData());
```

**NEW** SolverCG
```
  SolverCG(SolverControl &       cn,
           const MPI_Comm &      mpi_communicator,
           const AdditionalData &data = AdditionalData());
```

### CODE REVIEW

File: include/deal.II/lac/solver_cg.h [hunk 1]
```diff
@@ -411,6 +418,17 @@ SolverCG<VectorType>::SolverCG(SolverControl &cn, const AdditionalData &data)
 
 
 
+template <typename VectorType>
+SolverCG<VectorType>::SolverCG(SolverControl &       cn,
+                               const MPI_Comm &      mpi_communicator,
+                               const AdditionalData &data)
+  : SolverCG<VectorType>(cn, data)
+{
+  (void)mpi_communicator; // unused, for compatibility with PETSc
```
Comment (bangerth) [inline include/deal.II/lac/solver_cg.h:427]: If you just don't give the argument a name, you don't need to void it :-)

**Review by bangerth — COMMENTED** [review]
I don't particularly like this. We shouldn't be adding constructors we don't need to classes that don't need the information. A better question is why the PETSc class actually needs this. Couldn't it just take the communicator from the matrix or vectors we use?

**Comment by kronbichler** [comment]
> I don't particularly like this. We shouldn't be adding constructors we don't need to classes that don't need the information. A better question is why the PETSc class actually needs this. Couldn't it just take the communicator from the matrix or vectors we use?

I agree, we should try to avoid this if possible. I had a look at where the communicator gets used, and all places are like this one: https://github.com/dealii/dealii/blob/9216e16e2e52f1331993f6e3c0b8721dc9f9f116/source/lac/petsc_solver.cc#L71-L78
This means that we could ask the underlying vector class for the MPI communicator (or matrix/preconditioner classes). Or is there a case where it wouldn't work. (We have a comment a few lines above the code I referenced to, but I do not see this as excluding that path.)

**Comment by bangerth** [comment]
I'm working on it. Let's see whether I can get that done before my free hour is up!

**Comment by bangerth** [comment]
See #14438.

**Comment by masterleinad** [comment]
Superseded by https://github.com/dealii/dealii/pull/14438.

**Pull request closed without merging**